// kaiku/dist/kaiku.js
var _a;
/** @license Kaiku
 * kaiku.ts
 *
 * Copyright (c) 2021 Teemu Pääkkönen
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const CLASS_COMPONENT_FLAG = Symbol();
const IMMUTABLE_FLAG = Symbol();
const STATE_FLAG = Symbol();
const UNWRAP = Symbol();
const SVG_NAMESPACE_URI = "http://www.w3.org/2000/svg";
class Component {
  constructor(props) {
    this.props = {};
    this.state = void 0;
    this.props = props;
  }
  componentDidMount() {
  }
  componentWillUnmount() {
  }
}
_a = CLASS_COMPONENT_FLAG;
Component[_a] = true;
const EMPTY_OBJECT = false ? Object.freeze({}) : {};
function __assert(condition, message) {
  if (!Boolean(condition)) {
    throw new Error(message ?? "assert");
  }
}
const unionOfKeys = (a, b) => {
  const res = Object.keys(a);
  const bKeys = Object.keys(b);
  for (const key of bKeys) {
    if (!(key in a)) {
      res.push(key);
    }
  }
  return res;
};
const assert = false ? __assert : void 0;
let nextDependeeId = 0;
const updateDependee = (dependee) => {
  switch (dependee.tag_) {
    case 0:
    case 1: {
      updateComponentInstance(dependee);
      break;
    }
    case 5: {
      runEffect(dependee);
      break;
    }
    case 6:
    case 7: {
      runLazyUpdate(dependee);
      break;
    }
  }
};
let nextObjectId = 0;
let nextKeyId = 0;
const trackedDependencyStack = [];
const currentDependees = new Map();
const dependeeToKeys = new Map();
const keyToDependees = new Map();
let updatedDependencies = [];
const keyToId = new Map();
const getKeyId = (key) => {
  const existingKey = keyToId.get(key);
  if (existingKey) {
    return existingKey;
  }
  const id = ++nextKeyId * 67108864;
  keyToId.set(key, id);
  return id;
};
let deferredUpdateQueued = false;
const deferredUpdate = () => {
  deferredUpdateQueued = false;
  const updatedDependees = new Set();
  const keys = [...updatedDependencies];
  updatedDependencies = [];
  for (const key of keys) {
    const dependees = keyToDependees.get(key);
    if (!dependees)
      continue;
    for (const dependeeId of dependees) {
      if (updatedDependees.has(dependeeId)) {
        continue;
      }
      updatedDependees.add(dependeeId);
      const dependee = currentDependees.get(dependeeId);
      assert?.(dependee);
      updateDependee(dependee);
    }
  }
  if (updatedDependencies.length !== 0) {
    deferredUpdateQueued = true;
    if (false) {
      try {
        deferredUpdate();
      } catch (err) {
        if (err instanceof RangeError || typeof InternalError !== "undefined" && err instanceof InternalError) {
          throw new Error("useEffect loop detected! Make sure effects aren't continuously updating the same value.");
        } else {
          throw err;
        }
      }
    } else {
      deferredUpdate();
    }
  }
};
const trackedExecute = (dependee, fn, ...args) => {
  currentDependees.set(dependee.id_, dependee);
  const previousDependencies = dependeeToKeys.get(dependee.id_);
  if (previousDependencies) {
    for (const key of previousDependencies) {
      const dependees = keyToDependees.get(key);
      assert?.(dependees);
      dependees.delete(dependee.id_);
      if (dependees.size === 0) {
        keyToDependees.delete(key);
      }
    }
  }
  trackedDependencyStack.push(new Set());
  const result = fn(...args);
  const dependencies = trackedDependencyStack.pop();
  assert?.(dependencies);
  dependeeToKeys.set(dependee.id_, dependencies);
  for (const key of dependencies) {
    let dependencies2 = keyToDependees.get(key);
    if (!dependencies2) {
      dependencies2 = new Set();
      keyToDependees.set(key, dependencies2);
    }
    dependencies2.add(dependee.id_);
  }
  return result;
};
const removeDependencies = (dependee) => {
  currentDependees.delete(dependee.id_);
  const keys = dependeeToKeys.get(dependee.id_);
  if (keys) {
    for (const key of keys) {
      keyToDependees.get(key)?.delete(dependee.id_);
    }
    dependeeToKeys.delete(dependee.id_);
  }
};
const internalCreateState = (obj, shallow) => {
  const id = ++nextObjectId;
  const isArray = Array.isArray(obj);
  let initializing = true;
  const proxy = new Proxy(obj, {
    get(target, key) {
      if (key === UNWRAP)
        return target;
      if (key === STATE_FLAG)
        return true;
      if (trackedDependencyStack.length) {
        const dependencyKey = id + getKeyId(key);
        trackedDependencyStack[trackedDependencyStack.length - 1].add(dependencyKey);
      }
      const value = target[key];
      if (!isArray && typeof value === "function") {
        return value.bind(target);
      }
      return value;
    },
    deleteProperty(target, _key) {
      const key = _key;
      if (!initializing) {
        updatedDependencies.push(id);
        if (!deferredUpdateQueued) {
          deferredUpdateQueued = true;
          window.queueMicrotask(deferredUpdate);
        }
      }
      return delete target[key];
    },
    ownKeys(target) {
      if (trackedDependencyStack.length) {
        trackedDependencyStack[trackedDependencyStack.length - 1].add(id);
      }
      return Reflect.ownKeys(target);
    },
    has(target, key) {
      if (trackedDependencyStack.length) {
        trackedDependencyStack[trackedDependencyStack.length - 1].add(id);
      }
      return Reflect.has(target, key);
    },
    set(target, _key, value) {
      const key = _key;
      const isNewKeyForTarget = !(key in target);
      const isValueObject = typeof value === "object";
      const isValueNull = value === null;
      const isArrayLengthUpdate = isArray && key === "length";
      if (target[key] === value && !isArrayLengthUpdate && (isValueNull || !isValueObject || value[STATE_FLAG])) {
        return true;
      }
      if (!isValueNull && isValueObject && !value[STATE_FLAG] && !value[IMMUTABLE_FLAG] && !shallow) {
        target[key] = createState(value);
      } else {
        target[key] = value;
      }
      if (initializing) {
        return true;
      }
      if (isNewKeyForTarget) {
        updatedDependencies.push(id);
      }
      const dependencyKey = id + getKeyId(key);
      updatedDependencies.push(dependencyKey);
      if (!deferredUpdateQueued) {
        deferredUpdateQueued = true;
        window.queueMicrotask(deferredUpdate);
      }
      return true;
    }
  });
  if (!shallow) {
    for (const [key, value] of Object.entries(obj)) {
      if (typeof value === "object") {
        proxy[key] = value;
      }
    }
  }
  initializing = false;
  return proxy;
};
const createState = (obj) => internalCreateState(obj, false);
const createShallowState = (obj) => internalCreateState(obj, true);
const unwrap = (state) => {
  return state[UNWRAP];
};
const immutable = (obj) => {
  return new Proxy(obj, {
    get(target, _key) {
      const key = _key;
      if (key === IMMUTABLE_FLAG) {
        return true;
      }
      return target[key];
    }
  });
};
const effects = new Map();
const componentStates = new Map();
const componentStateIndexStack = [];
const componentIdStack = [];
const componentsThatHaveUpdatedAtLeastOnce = new Set();
const startHookTracking = (componentId) => {
  componentIdStack.push(componentId);
  componentStateIndexStack.push(0);
};
const stopHookTracking = () => {
  const refIndex = componentStateIndexStack.pop();
  assert?.(typeof refIndex !== "undefined");
  const componentId = componentIdStack.pop();
  assert?.(typeof componentId !== "undefined");
  componentsThatHaveUpdatedAtLeastOnce.add(componentId);
};
const destroyHooks = (componentId) => {
  componentsThatHaveUpdatedAtLeastOnce.delete(componentId);
  componentStates.delete(componentId);
  const componentEffects = effects.get(componentId);
  if (componentEffects) {
    for (const effect of componentEffects) {
      removeDependencies(effect);
      effect.unsubscribe_?.();
    }
    effects.delete(componentId);
  }
};
const runEffect = (effect) => {
  effect.unsubscribe_?.();
  effect.unsubscribe_ = trackedExecute(effect, effect.fn);
};
const useEffect = (fn) => {
  const componentId = componentIdStack[componentIdStack.length - 1];
  if (componentId !== void 0 && componentsThatHaveUpdatedAtLeastOnce.has(componentId)) {
    return;
  }
  const effect = {
    id_: ++nextDependeeId,
    tag_: 5,
    fn
  };
  runEffect(effect);
  if (!componentId) {
    return;
  }
  let componentEffects = effects.get(componentId);
  if (!componentEffects) {
    componentEffects = [];
    effects.set(componentId, componentEffects);
  }
  assert?.(componentEffects);
  componentEffects.push(effect);
};
const internalUseState = (initialState, shallow) => {
  const componentId = componentIdStack[componentIdStack.length - 1];
  const componentStateIndex = componentStateIndexStack[componentStateIndexStack.length - 1]++;
  assert?.(typeof componentId !== "undefined");
  let states = componentStates.get(componentId);
  if (!states) {
    states = [];
    componentStates.set(componentId, states);
  }
  if (states.length > componentStateIndex) {
    return states[componentStateIndex];
  }
  const componentState = internalCreateState(initialState, shallow);
  states.push(componentState);
  return componentState;
};
const useState = (initialState) => internalUseState(initialState, false);
const useShallowState = (initialState) => internalUseState(initialState, true);
const useRef = (initialValue) => internalUseState({ current: initialValue }, true);
const childToDescriptor = (child) => {
  if (child === null || child === void 0 || typeof child === "boolean") {
    return { tag_: 3, value_: "" };
  }
  if (typeof child === "string" || typeof child === "number") {
    return { tag_: 3, value_: child };
  }
  if (Array.isArray(child)) {
    return createFragmentDescriptor(EMPTY_OBJECT, child);
  }
  if (typeof child === "function") {
    return h(child, EMPTY_OBJECT);
  }
  return child;
};
const reuseNodeInstance = (instance, descriptor) => {
  const isSameFunctionComponent = instance.tag_ === 0 && descriptor.tag_ === 0 && instance.func === descriptor.func;
  const isSameClassComponent = instance.tag_ === 1 && descriptor.tag_ === 1 && instance.instance instanceof descriptor.class_;
  if (isSameFunctionComponent || isSameClassComponent) {
    assert?.(instance.tag_ === 0 || instance.tag_ === 1);
    assert?.(descriptor.tag_ === 0 || descriptor.tag_ === 1);
    const keys = unionOfKeys(descriptor.props_, instance.props_);
    for (const key of keys) {
      instance.props_[key] = descriptor.props_[key];
    }
    return true;
  }
  if (instance.tag_ === 2 && descriptor.tag_ === 2 && instance.tagName_ === descriptor.tagName_) {
    updateHtmlElementInstance(instance, descriptor.props_, descriptor.children_);
    return true;
  }
  if (instance.tag_ === 4 && descriptor.tag_ === 4) {
    updateFragmentInstance(instance, descriptor.children_);
    return true;
  }
  if (instance.tag_ === 3 && descriptor.tag_ === 3) {
    if (instance.element_.data !== String(descriptor.value_)) {
      instance.element_.data = descriptor.value_;
    }
    return true;
  }
  return false;
};
const getKeyOfNodeDescriptor = (descriptor) => {
  if (descriptor.tag_ === 3) {
    return null;
  }
  if (typeof descriptor.props_?.key !== "undefined") {
    return descriptor.props_?.key;
  }
  return null;
};
const createClassComponentDescriptor = (class_, props_) => ({
  tag_: 1,
  class_,
  props_
});
const createClassComponentInstance = (descriptor, parentElement) => {
  const id = ++nextDependeeId;
  startHookTracking(id);
  const classInstance = new descriptor.class_(descriptor.props_);
  stopHookTracking();
  classInstance.render = classInstance.render.bind(classInstance);
  if (typeof classInstance.state !== "undefined") {
    classInstance.state = internalCreateState(classInstance.state, false);
  }
  const instance = {
    id_: id,
    tag_: 1,
    hasMounted: false,
    instance: classInstance,
    props_: internalCreateState(descriptor.props_, true),
    parentElement_: parentElement
  };
  updateComponentInstance(instance);
  return instance;
};
const createFunctionComponentDescriptor = (func, props_) => ({
  tag_: 0,
  func,
  props_
});
const createFunctionComponentInstance = (descriptor, parentElement) => {
  const instance = {
    id_: ++nextDependeeId,
    tag_: 0,
    hasMounted: false,
    func: descriptor.func,
    props_: internalCreateState(descriptor.props_, true),
    parentElement_: parentElement
  };
  updateComponentInstance(instance);
  return instance;
};
const updateComponentInstance = (instance, props = instance.props_) => {
  let childDescriptor;
  if (instance.tag_ === 0) {
    startHookTracking(instance.id_);
    childDescriptor = trackedExecute(instance, instance.func, instance.props_);
    stopHookTracking();
  } else {
    instance.instance.props = props;
    childDescriptor = trackedExecute(instance, instance.instance.render, props);
  }
  if (!childDescriptor) {
    if (instance.child) {
      unmountNodeInstance(instance.child);
    }
    delete instance.child;
    return;
  }
  if (!instance.child) {
    instance.child = createNodeInstance(childDescriptor, instance.parentElement_);
  } else {
    const wasReused = reuseNodeInstance(instance.child, childDescriptor);
    if (!wasReused) {
      unmountNodeInstance(instance.child);
      const newChild = createNodeInstance(childDescriptor, instance.parentElement_);
      instance.child = newChild;
    }
  }
  if (instance.hasMounted) {
    mountNodeInstance(instance, instance.parentElement_, instance.nextSibling_);
  }
};
const Fragment = false ? () => assert(false, "Fragment should not be called explicitly") : () => {
};
const createFragmentDescriptor = (props_, children) => {
  return {
    tag_: 4,
    children_: children,
    props_
  };
};
const createFragmentInstance = (descriptor, parentElement) => {
  const instance = {
    tag_: 4,
    props_: descriptor.props_,
    hasMounted: false,
    parentElement_: parentElement,
    childMap: new Map(),
    children_: []
  };
  updateFragmentInstance(instance, descriptor.children_);
  return instance;
};
const updateFragmentInstance = (instance, children) => {
  const childDescriptors = children.map(childToDescriptor);
  instance.children_ = [];
  const nextkeys = new Set();
  for (let i = childDescriptors.length - 1; i >= 0; i--) {
    const descriptor = childDescriptors[i];
    assert?.(descriptor);
    const descriptorKey = getKeyOfNodeDescriptor(descriptor);
    const key = descriptorKey === null ? i : descriptorKey;
    if (false) {
      if (nextkeys.has(key)) {
        throw new Error("Duplicate key detected! Make sure no two sibling elements have the same key.");
      }
    }
    nextkeys.add(key);
    const existingChild = instance.childMap.get(key);
    const wasReused = existingChild && reuseNodeInstance(existingChild, descriptor);
    if (wasReused) {
      assert?.(existingChild);
      instance.children_.push(existingChild);
    } else {
      if (existingChild) {
        unmountNodeInstance(existingChild);
      }
      const node = createNodeInstance(descriptor, instance.parentElement_);
      instance.children_.push(node);
      instance.childMap.set(key, node);
    }
  }
  for (const [key, child] of instance.childMap) {
    if (!nextkeys.has(key)) {
      unmountNodeInstance(child);
      instance.childMap.delete(key);
    }
  }
};
const createTextInstance = (descriptor, parentElement) => {
  return {
    tag_: 3,
    element_: document.createTextNode(descriptor.value_),
    hasMounted: false,
    parentElement_: parentElement
  };
};
const updateElementProperty = (element, property, value) => {
  assert?.(property !== "style");
  switch (property) {
    case "value": {
      ;
      element.value = value;
      break;
    }
    case "checked": {
      if (Boolean(value)) {
        element.setAttribute(property, value);
      } else {
        element.removeAttribute(property);
      }
      ;
      element.checked = Boolean(value);
      break;
    }
    case "class":
    case "className": {
      ;
      element.className = stringifyClassNames(value ?? "");
      break;
    }
    case "html": {
      ;
      element.innerHTML = value;
      break;
    }
    default: {
      if (!Boolean(value) && value !== 0) {
        element.removeAttribute(property);
      } else {
        element.setAttribute(property, value);
      }
      break;
    }
  }
};
const updateElementStyle = (element, property, value) => {
  element.style[property] = value;
};
const runLazyUpdate = (lazyUpdate) => {
  const value = trackedExecute(lazyUpdate, lazyUpdate.callback);
  if (value !== lazyUpdate.previousValue) {
    lazyUpdate.previousValue = value;
    if (lazyUpdate.tag_ === 6) {
      updateElementProperty(lazyUpdate.element_, lazyUpdate.property, value);
    } else {
      updateElementStyle(lazyUpdate.element_, lazyUpdate.property, value);
    }
  }
};
const registerLazyUpdate = (instance, property, callback, type) => {
  instance.lazyUpdates ??= [];
  let lazyUpdate;
  for (const existingUpdate of instance.lazyUpdates) {
    if (existingUpdate.tag_ === type && existingUpdate.property === property) {
      lazyUpdate = existingUpdate;
      break;
    }
  }
  if (lazyUpdate) {
    lazyUpdate.callback = callback;
  } else {
    lazyUpdate = {
      id_: ++nextDependeeId,
      tag_: type,
      element_: instance.element_,
      property,
      callback,
      previousValue: void 0
    };
    instance.lazyUpdates.push(lazyUpdate);
  }
  runLazyUpdate(lazyUpdate);
};
const stringifyClassNames = (names) => {
  if (names === null || typeof names === "undefined") {
    return "";
  }
  if (typeof names === "string") {
    return names;
  }
  let className = "";
  if (Array.isArray(names)) {
    for (const name of names) {
      className += stringifyClassNames(name) + " ";
    }
    return className.trim();
  }
  const keys = Object.keys(names);
  for (const key of keys) {
    if (names[key])
      className += key + " ";
  }
  return className.trim();
};
const createHtmlElementDescriptor = (tagName_, props_, children_) => ({
  tag_: 2,
  tagName_,
  props_,
  children_
});
const createHtmlElementInstance = (descriptor, parentElement) => {
  const useSvgNs = descriptor.tagName_ === "svg" || parentElement.namespaceURI === SVG_NAMESPACE_URI;
  const element_ = useSvgNs ? document.createElementNS(SVG_NAMESPACE_URI, descriptor.tagName_) : document.createElement(descriptor.tagName_);
  const instance = {
    tag_: 2,
    tagName_: descriptor.tagName_,
    hasMounted: false,
    element_,
    parentElement_: parentElement,
    props_: EMPTY_OBJECT
  };
  updateHtmlElementInstance(instance, descriptor.props_, descriptor.children_);
  return instance;
};
const updateHtmlElementInstance = (instance, nextProps, children) => {
  const keys = unionOfKeys(nextProps, instance.props_);
  const properties = unionOfKeys(nextProps.style ?? EMPTY_OBJECT, instance.props_.style ?? EMPTY_OBJECT);
  for (const property of properties) {
    const prevValue = instance.props_.style?.[property];
    const value = nextProps.style?.[property];
    if (prevValue !== value) {
      if (typeof value === "function") {
        registerLazyUpdate(instance, property, value, 7);
      } else {
        updateElementStyle(instance.element_, property, value);
      }
    }
  }
  for (const key of keys) {
    if (key === "style")
      continue;
    if (key === "key")
      continue;
    if (instance.props_[key] === nextProps[key])
      continue;
    if (key === "ref") {
      if (key in nextProps) {
        nextProps[key].current = instance.element_;
      } else {
        instance.props_[key].current = void 0;
      }
      continue;
    }
    const isListener = key[0] === "o" && key[1] === "n";
    if (isListener) {
      const eventName = key.substring(2).toLowerCase();
      if (!instance.eventListener) {
        instance.eventHandlers = {};
        instance.eventListener = (evt) => {
          const handlers = instance.eventHandlers;
          assert?.(handlers, "instance.eventHandlers record not initialized before event call");
          const handler = handlers[evt.type];
          assert?.(handler, `handler for event type '${evt.type}' not present in instance.eventHandlers`);
          return handler(evt);
        };
      }
      assert?.(instance.eventHandlers);
      assert?.(instance.eventListener);
      if (typeof nextProps[key] === "function") {
        if (!(eventName in instance.eventHandlers)) {
          instance.element_.addEventListener(eventName, instance.eventListener);
        }
        instance.eventHandlers[eventName] = nextProps[key];
      } else {
        instance.element_.removeEventListener(eventName, instance.eventListener);
        delete instance.eventHandlers[eventName];
      }
    } else {
      if (typeof nextProps[key] === "function") {
        registerLazyUpdate(instance, key, nextProps[key], 6);
      } else {
        updateElementProperty(instance.element_, key, nextProps[key]);
      }
    }
  }
  instance.props_ = nextProps;
  if (children.length === 0) {
    if (instance.children_) {
      unmountNodeInstance(instance.children_);
    }
    delete instance.children_;
  } else {
    if (instance.children_) {
      reuseNodeInstance(instance.children_, createFragmentDescriptor(EMPTY_OBJECT, children));
    } else {
      instance.children_ = createFragmentInstance(createFragmentDescriptor(EMPTY_OBJECT, children), instance.element_);
    }
  }
};
const NODE_CREATORS = {
  [1]: createClassComponentInstance,
  [0]: createFunctionComponentInstance,
  [2]: createHtmlElementInstance,
  [4]: createFragmentInstance,
  [3]: createTextInstance
};
const createNodeInstance = (descriptor, parentElement) => NODE_CREATORS[descriptor.tag_](descriptor, parentElement);
const getNextSiblingElement = (instance) => {
  switch (instance.tag_) {
    case 0:
    case 1: {
      if (instance.child) {
        return getNextSiblingElement(instance.child);
      }
      if (instance.nextSibling_) {
        return getNextSiblingElement(instance.nextSibling_);
      }
      return null;
    }
    case 2:
    case 3:
      return instance.element_;
    case 4: {
      if (instance.children_.length !== 0) {
        return getNextSiblingElement(instance.children_[instance.children_.length - 1]);
      }
      if (instance.nextSibling_) {
        return getNextSiblingElement(instance.nextSibling_);
      }
      return null;
    }
  }
};
const mountNodeInstance = (instance, parentElement, nextSibling) => {
  assert?.(instance);
  assert?.(parentElement);
  switch (instance.tag_) {
    case 0:
    case 1: {
      if (instance.child) {
        mountNodeInstance(instance.child, parentElement, nextSibling);
      }
      if (instance.tag_ === 1 && !instance.hasMounted) {
        instance.instance.componentDidMount();
      }
      instance.nextSibling_ = nextSibling;
      break;
    }
    case 2:
    case 3: {
      if (instance.tag_ === 2 && instance.children_) {
        mountNodeInstance(instance.children_, instance.element_);
      }
      if (instance.hasMounted && instance.nextSibling_ === nextSibling) {
        break;
      }
      instance.nextSibling_ = nextSibling;
      const refElement = nextSibling && getNextSiblingElement(nextSibling);
      parentElement.insertBefore(instance.element_, refElement);
      break;
    }
    case 4: {
      instance.nextSibling_ = nextSibling;
      for (const child of instance.children_) {
        mountNodeInstance(child, parentElement, nextSibling);
        nextSibling = child;
      }
      break;
    }
  }
  instance.hasMounted = true;
};
const unmountNodeInstance = (instance) => {
  switch (instance.tag_) {
    case 0:
    case 1: {
      destroyHooks(instance.id_);
      removeDependencies(instance);
      if (instance.tag_ === 1) {
        instance.instance.componentWillUnmount();
      }
      if (instance.child) {
        unmountNodeInstance(instance.child);
      }
      break;
    }
    case 2: {
      if (typeof instance.props_.ref !== "undefined") {
        instance.props_.ref.current = void 0;
      }
      instance.lazyUpdates?.forEach(removeDependencies);
      assert?.(instance.parentElement_);
      instance.parentElement_.removeChild(instance.element_);
      if (instance.children_) {
        instance.children_.children_.forEach(unmountNodeInstance);
      }
      break;
    }
    case 3: {
      assert?.(instance.parentElement_);
      instance.parentElement_.removeChild(instance.element_);
      break;
    }
    case 4: {
      instance.children_.forEach(unmountNodeInstance);
      break;
    }
  }
};
function h(type, props, ...children) {
  if (typeof type === "string") {
    return createHtmlElementDescriptor(type, props ?? EMPTY_OBJECT, children);
  }
  if (type[CLASS_COMPONENT_FLAG]) {
    return createClassComponentDescriptor(type, {
      ...props ?? EMPTY_OBJECT,
      children: children.length === 0 ? void 0 : children
    });
  }
  if (type === Fragment) {
    return createFragmentDescriptor(props ?? EMPTY_OBJECT, children);
  }
  return createFunctionComponentDescriptor(type, {
    ...props ?? EMPTY_OBJECT,
    children: children.length === 0 ? void 0 : children
  });
}
function jsx(type, props, key) {
  let children = props?.children;
  children = children !== void 0 ? Array.isArray(children) ? children : [children] : void 0;
  const propsCopy = {
    ...props,
    children,
    key
  };
  if (typeof type === "string") {
    delete propsCopy.children;
    return createHtmlElementDescriptor(type, propsCopy, children ?? []);
  }
  if (type[CLASS_COMPONENT_FLAG]) {
    return createClassComponentDescriptor(type, propsCopy);
  }
  if (type === Fragment) {
    delete propsCopy.children;
    return createFragmentDescriptor(propsCopy, children ?? []);
  }
  return createFunctionComponentDescriptor(type, propsCopy);
}
const render = (descriptor, element) => mountNodeInstance(createNodeInstance(descriptor, element), element);
export {
  Component,
  Fragment,
  h,
  jsx,
  render,
  useState,
  useShallowState,
  useRef,
  useEffect,
  createState,
  createShallowState,
  immutable,
  unwrap
};
